citro3d/examples/cube.rs

```diff
@@ -0,0 +1,250 @@
+    let mut hid = Hid::new().expect("Couldn't obtain HID controller");
+    let apt = Apt::new().expect("Couldn't obtain APT controller");
+
+    let mut instance = citro3d::Instance::new().expect("failed to initialize Citro3D");
+
+    let top_screen = TopScreen3D::from(&gfx.top_screen);
+
+    let (mut top_left, mut top_right) = top_screen.split_mut();
+
+    let RawFrameBuffer { width, height, .. } = top_left.raw_framebuffer();
+    let mut top_left_target =
+        render::Target::new(width, height, top_left, None).expect("failed to create render target");
+
+    let RawFrameBuffer { width, height, .. } = top_right.raw_framebuffer();
+    let mut top_right_target = render::Target::new(width, height, top_right, None)
+        .expect("failed to create render target");
+
+    let mut bottom_screen = gfx.bottom_screen.borrow_mut();
+    let RawFrameBuffer { width, height, .. } = bottom_screen.raw_framebuffer();
+
+    let mut bottom_target = render::Target::new(width, height, bottom_screen, None)
+        .expect("failed to create bottom screen render target");
+
+    let shader = shader::Library::from_bytes(SHADER_BYTES).unwrap();
+    let vertex_shader = shader.get(0).unwrap();
+
+    let program = shader::Program::new(vertex_shader).unwrap();
+    instance.bind_program(&program);
+    let mut vbo_data = Vec::with_capacity_in(VERTS.len(), ctru::linear::LinearAllocator);
+    for vert in VERTS.iter().enumerate().map(|(i, v)| Vertex {
+        pos: Vec3 {
+            x: v[0],
+            y: v[1],
+            z: v[2],
+        },
+        color: Vec3::new(1.0, 0.7, 0.5),
+    }) {
+        vbo_data.push(vert);
+    }
+
+    let mut buf_info = buffer::Info::new();
+    let (attr_info, vbo_data) = prepare_vbos(&mut buf_info, &vbo_data);
+
+    // Configure the first fragment shading substage to just pass through the vertex color
+    // See https://www.opengl.org/sdk/docs/man2/xhtml/glTexEnv.xml for more insight
+    let stage0 = texenv::Stage::new(0).unwrap();
+    instance
+        .texenv(stage0)
+        .src(texenv::Mode::BOTH, texenv::Source::PrimaryColor, None, None)
+        .func(texenv::Mode::BOTH, texenv::CombineFunc::Replace);
+
+    let projection_uniform_idx = program.get_uniform("projection").unwrap();
+    let camera_transform = Matrix4::looking_at(
+        FVec3::new(1.8, 1.8, 1.8),
+        FVec3::new(0.0, 0.0, 0.0),
+        FVec3::new(0.0, 1.0, 0.0),
+        CoordinateOrientation::RightHanded,
+    );
+    let indecies_a = [
+        0, 3, 1, 1, 3, 2, // triangles making up the top (+y) facing side.
+        4, 5, 7, 5, 6, 7, // bottom (-y)
+        8, 11, 9, 9, 11, 10, // right (+x)
+        12, 13, 15, 13, 14, 15, // left (-x)
+        16, 19, 17, 17, 19, 18, // back (+z)
+        20, 21, 23, 21, 22, 23, // forward (-z)
+    ];
+    let mut indecies = Vec::with_capacity_in(indecies_a.len(), ctru::linear::LinearAllocator);
+    indecies.extend_from_slice(&indecies_a);
```

Comment:
Maybe to enforce lifetimes for index buffers we could have something like `buffer::Info::index_buffer(&[usize])`. I think if we had that it could resolve at least partially some of the lifetime issues, since `Slice` already captures the lifetime of the vertex buffer. Then I think the code would look something like this?

```rust
let (attr_info, _vbo_data) = prepare_vbos(&mut buf_info, &vbo_data);

let mut index_buffer = buf_info.index_buffer();
// This possibly also checks that all  the new indices are valid for the vertex data:
index_buffer.extend_from_slice(&indices_a);
```

Edit: actually, I guess we'd also need to make sure the vertex buffer was added at some point, so maybe this would be a method on `buffer::Slice` instead, or a new `buffer::Info::add_indexed(vbo_data, attrib_info, indices)` or something. I'm not sure what the best API would be but maybe something along those lines